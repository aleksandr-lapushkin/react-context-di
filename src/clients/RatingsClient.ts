import { AxiosInstance } from "axios";

export interface RatingsClient {
    getRatingForUser(id: string): Promise<number>
}

export class AxiosRatingsClient implements RatingsClient {
    private axiosInstance: AxiosInstance;

    constructor(axiosInstance: AxiosInstance) {
        this.axiosInstance = axiosInstance
    }

    async getRatingForUser(id: string): Promise<number> {
        const result = await this.axiosInstance.get<{rating: number}>(`/api/users/${id}/rating`)

        if (result.data) {
            return result.data.rating
        }

        throw new Error("Unable to fetch Rating")
    }

}