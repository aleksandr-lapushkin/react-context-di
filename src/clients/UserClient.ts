import { AxiosInstance } from "axios";
import {ApiPaths} from "../constants/ApiPaths";

export interface User {
    id: string
    name: string
    fullName: string
}

export interface UserClient {
    getUser(id: string): Promise<User>
    getUsers(): Promise<User[]>
    getUserSafe(id: string): Promise<Result<User, ResultCode>>
}

export class AxiosUserClient implements UserClient {
    private axiosInstance: AxiosInstance;

    constructor(axiosInstance: AxiosInstance) {
        this.axiosInstance = axiosInstance;
    }

    async getUser(id: string): Promise<User> {
        try {
            const result = await this.axiosInstance.get<User>(ApiPaths.users.read(id))

            if (result.data) {
                return result.data;
            }
        } catch (err) {
            if (err?.response?.status === 404) {
                throw new Error("Not Found")
            }
            throw err
        }

        throw new Error("Failed to fetch User")
    }

    async getUsers(): Promise<User[]> {
        const result = await this.axiosInstance.get<{users: User[]}>(ApiPaths.users.list)

        if (result.data) {
            return result.data.users;
        }

        throw new Error("Failed to fetch Users")
    }

    async getUserSafe(id: string): Promise<Result<User, ResultCode>> {
        try {
            const result = await this.axiosInstance.get<User>(ApiPaths.users.read(id))
            return new SuccessImpl<User>(result.data)

        } catch (err) {
            if (err?.response?.status === 404) {
                return new ServerFailureImpl<ResultCode>(ResultCode.NOT_FOUND)
            }
            if (err?.response?.data) {
                return new ServerFailureImpl(ResultCode.UNKNOWN)
            }
            return new RequestFailureImpl(err.toString())
        }
    }

}












//Bonus stuff
export enum ResultCode {
    OK,
    NOT_FOUND,
    UNKNOWN
}
export enum ResultType {
    SUCCESS,
    SERVER_FAILURE,
    REQUEST_FAILURE
}
interface Success<T> extends Fold<T, any>{
    data: T,
    type: ResultType.SUCCESS
}
interface SuccessFolder<T, O> {
    (data: T): O
}
interface ServerFailure<E> extends Fold<any, E>{
    error: E,
    type: ResultType.SERVER_FAILURE
}
interface ServerFailureFolder<E, O> {
    (data: E): O
}
interface RequestFailure extends Fold<any, any> {
    message?: string,
    type: ResultType.REQUEST_FAILURE
}
interface RequestFailureFolder<O> {
    (data?: string): O
}
interface Fold<T, E> {
    fold: <O>(s: SuccessFolder<T, O>, sf: ServerFailureFolder<E, O>, rf: RequestFailureFolder<O>) => O
}
export type Result<T, E> = (Success<T> | ServerFailure<E> | RequestFailure) & Fold<T, E>
export class SuccessImpl<T> implements Success<T> {
    data: T;
    type: ResultType.SUCCESS;

    constructor(data: T) {
        this.type = ResultType.SUCCESS
        this.data = data
    }

    public fold<O>(s: SuccessFolder<T, O>, sf: ServerFailureFolder<any, O>, rf: RequestFailureFolder<O>): O {
        return s(this.data)
    }
}
export class ServerFailureImpl<E> implements ServerFailure<E> {
    error: E;
    type: ResultType.SERVER_FAILURE;
    constructor(error: E) {
        this.type = ResultType.SERVER_FAILURE;
        this.error = error;
    }
    fold<O>(s: SuccessFolder<any, O>, sf: ServerFailureFolder<E, O>, rf: RequestFailureFolder<O>): O {
        return sf(this.error)
    }
}
export class RequestFailureImpl implements RequestFailure {
    message?: string;
    type: ResultType.REQUEST_FAILURE;
    constructor(message?: string) {
        this.message = message;
        this.type = ResultType.REQUEST_FAILURE;
    }
    fold<O>(s: SuccessFolder<any, O>, sf: ServerFailureFolder<any, O>, rf: RequestFailureFolder<O>): O {
        return rf(this.message)
    }

}