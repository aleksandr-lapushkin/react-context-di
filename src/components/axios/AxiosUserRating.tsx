import React, {PropsWithChildren, useEffect, useState} from "react";
import axios from "axios"
import {UserRating} from "../display/UserRating";

export interface AxiosUserRatingProps {
    userId: string
}

export const AxiosUserRating: React.FC<PropsWithChildren<AxiosUserRatingProps>> = ({userId, children}) => {
    const [rating, setRating] = useState<number>()
    const [error, setError] = useState<string>()

    useEffect(() => {
        const loadRating = async () => {
            try {
                const result = await axios.get<{rating: number}>(`/api/users/${userId}/rating`)

                setRating(result.data.rating)
            } catch (err) {
                setError(err.response.data.message)
            }

        }
        loadRating()
    })

    return <UserRating error={error} rating={rating}>{children}</UserRating>
}