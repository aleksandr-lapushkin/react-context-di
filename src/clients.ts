import {AxiosUserClient, UserClient} from "./clients/UserClient";
import axios from "axios";
import {AxiosRatingsClient, RatingsClient} from "./clients/RatingsClient";

let axiosInstance = axios.create();
export const usersClient: UserClient = new AxiosUserClient(axiosInstance)
export const ratingsClient: RatingsClient = new AxiosRatingsClient(axiosInstance)