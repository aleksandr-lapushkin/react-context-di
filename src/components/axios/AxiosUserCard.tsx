import * as React from "react";
import {UserCard} from "../display/UserCard";
import axios from "axios";
import {User} from "../../clients/UserClient";
import {useEffect, useState} from "react";
import {UserAvatar} from "../display/UserAvatar";
import {AxiosUserRating} from "./AxiosUserRating";

export interface AxiosUserCardProps {
    id: string
}

export const AxiosUserCard: React.FC<AxiosUserCardProps> = ({id}) => {
    const [user, setUser] = useState<User>();

    useEffect(() => {
        const loadData = async () => {
            const fetchResult = await axios.get<User>(`/api/users/${id}`)
            if (fetchResult.data) {
                setUser(fetchResult.data)
            }
        }
        loadData()
    }, [id])

    return <UserCard user={user} titleRenderer={u => <div><AxiosUserRating userId={u.id}><UserAvatar/></AxiosUserRating><span style={{margin: "16px"}}>{u.name}</span></div>}/>
}