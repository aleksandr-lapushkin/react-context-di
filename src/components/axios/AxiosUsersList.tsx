import React, {useEffect, useState} from "react";
import {User} from "../../clients/UserClient";
import axios from "axios"
import {UsersList} from "../display/UsersList";

export const AxiosUsersList: React.FC = () => {
    const [users, setUsers] = useState<User[]>()
    const [error, setError] = useState<string>()
    useEffect(() => {
        const loadData = async () => {
            try {
                const fetchResult = await axios.get<{users: User[]}>("/api/users")
                if (fetchResult.data) {
                    setUsers(fetchResult.data.users)
                }
            } catch (err) {
                setError(err.response.data.message)
            }
        }
        loadData()
    }, [])

    return (
        <UsersList users={users} errorMessage={error}/>
    )
}