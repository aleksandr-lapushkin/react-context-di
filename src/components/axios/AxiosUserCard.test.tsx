import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import {User} from "../../clients/UserClient";
import {render} from "@testing-library/react";
import {StaticRouter} from "react-router";
import React from "react";
import {AxiosUserCard} from "./AxiosUserCard";

describe("Axios-backed UserCard", () => {
    const mock = new MockAdapter(axios);
    beforeEach(() => {
        mock.reset()
    })
    it("Can show a User's details", async () => {
        const user: User = {id: "asd", name: "someone", fullName: "Some One"}

        mock.onGet("/api/users/asd").reply(200, user)
        mock.onGet("/api/users/asd/rating").reply(200, {rating: 42})

        const element = render(
            <StaticRouter>
                <AxiosUserCard id={"asd"} />
            </StaticRouter>
        )

        let item = await element.findByText("someone");
        expect(item).toBeInTheDocument()
        let rating = await element.findByTitle("42");
        expect(rating).toBeInTheDocument()
        expect(element.container).toMatchSnapshot()
    })
})