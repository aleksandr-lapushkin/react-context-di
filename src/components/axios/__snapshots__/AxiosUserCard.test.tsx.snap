// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axios-backed UserCard Can show a User's details 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <div>
            <div>
              <span
                class="ant-badge"
              >
                <span
                  class="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image"
                >
                  <img
                    src="https://via.placeholder.com/150"
                  />
                </span>
                <sup
                  class="ant-scroll-number ant-badge-count ant-badge-multiple-words"
                  data-show="true"
                  title="42"
                >
                  <span
                    class="ant-scroll-number-only"
                    style="transform: translateY(-400%);"
                  >
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class="current"
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                  </span>
                  <span
                    class="ant-scroll-number-only"
                    style="transform: translateY(-200%);"
                  >
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class="current"
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                    <p
                      class=""
                    >
                      0
                    </p>
                    <p
                      class=""
                    >
                      1
                    </p>
                    <p
                      class=""
                    >
                      2
                    </p>
                    <p
                      class=""
                    >
                      3
                    </p>
                    <p
                      class=""
                    >
                      4
                    </p>
                    <p
                      class=""
                    >
                      5
                    </p>
                    <p
                      class=""
                    >
                      6
                    </p>
                    <p
                      class=""
                    >
                      7
                    </p>
                    <p
                      class=""
                    >
                      8
                    </p>
                    <p
                      class=""
                    >
                      9
                    </p>
                  </span>
                </sup>
              </span>
              <span
                style="margin: 16px;"
              >
                someone
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div>
        <div>
          ID: 
          asd
        </div>
        <div>
          Full Name: 
          Some One
        </div>
      </div>
    </div>
  </div>
</div>
`;
